test/unit/raft.c

```diff
@@ -1893,10 +1893,88 @@ raft_test_split_vote(void)
 	raft_finish_test();
 }
 
+static void
+raft_test_pre_vote(void)
+{
+	raft_start_test(13);
+	struct raft_node node;
+	raft_node_create(&node);
+
+	/* Check leader_idle calculations. */
+	is(raft_node_send_leader(&node,
+				 2 /* Term. */,
+				 2 /* Source. */
+	), 0, "leader notification");
+
+	is(raft_leader_idle(&node.raft), 0, "leader just appeared");
+
+	raft_run_for(node.cfg_death_timeout / 2);
+	is(raft_leader_idle(&node.raft), node.cfg_death_timeout / 2,
+	   "leader_idle increased");
+
+	raft_node_send_heartbeat(&node, 2);
+
+	is(raft_leader_idle(&node.raft), 0, "heartbeat resets idle counter");
+
+	raft_node_cfg_is_candidate(&node, false);
+
+	ok(raft_ev_is_active(&node.raft.timer), "voter tracks leader death");
+
+	raft_run_for(2 * node.cfg_death_timeout);
+
+	ok(raft_node_check_full_state(&node,
+				      RAFT_STATE_FOLLOWER /* State. */,
+				      2 /* Leader. */,
+				      2 /* Term. */,
+				      0 /* Vote. */,
+				      2 /* Volatile term. */,
+				      0 /* Volatile vote. */,
+				      "{0: 1}" /* Vclock. */
+	), "leader still remembered");
+
+	is(raft_leader_idle(&node.raft), 2 * node.cfg_death_timeout,
+	   "idle increased");
+
+	raft_node_send_heartbeat(&node, 2);
+	is(raft_leader_idle(&node.raft), 0, "heartbeat resets idle");
+	ok(raft_ev_is_active(&node.raft.timer), "heartbeat restarts timer");
+
+	raft_node_cfg_is_candidate(&node, true);
+
+	raft_node_notify_leader_seen(&node, NODE_3);
```

Comment:
Maybe just pass plain numbers like 1, 2, etc and plainly call `raft_notify_leader_seen()` internally? This `node.witness_map` is not used anyway. You save it in the first line of `raft_node_notify_leader_seen` but it is not accessed out of that function anywhere.

Reply:
Agree. Done.